### PR TITLE
Preliminary Videos

### PR DESCRIPTION
**Pull request by EricBAndrews**
Depends on https://github.com/mlemgroup/MlemMiddleware/pull/63

# Checklist
- [x] I have read CONTRIBUTING.md
- [x] I have described what this PR contains
- [x] If this PR alters the UI, I have attached pictures/videos
- [x] This PR addresses one or more open issues that were assigned to me:
      - progress towards #1306 

# Pull Request Information

This PR adds preliminary video handling, supporting the following formats:
- .webp
- .gif
- .mp4
- .mov
- .m4v

.mp4, .mov, and .m4v are all handed natively by NukeVideo.

.gif is handled by Nuke via Gifu; `GifView` provides a nice SwiftUI wrapper around the underlying Gifu UIKit component.

.webp is the odd format out--Nuke does not have any actively maintained support, so I have brought in a pair of SDWebImage libraries to handle it instead:
- SDWebImageSwiftUI gives us `AnimatedImage`, which can render animated images
- SDWebImageWebPCoder is a plugin that allows decoding of .webp for consumption by `AnimatedImage`

## Image Loading and Rendering

`ImageLoader` has been reworked into `MediaLoader`. Because different media types need to be handled by different views, its simple `uiImage` has been replaced by `mediaType`, an enum with associated values containing the assets needed to render both the still thumbnail and, if applicable, the animated media.

The new `MediaView` handles rendering media; it takes in a `MediaType` and renders the content. It takes a boolean `playing` that determines whether animated media should be playing or not. For performance, it only renders the animated media on-demand; until `playing` is toggled on, only the still image is rendered.

`DynamicMediaView` (previously `DynamicImageView`) is now additionally responsible for handling the playing state of its underlying `MediaView`. It renders a play button on top of animated media; if the button is tapped, the media will animate, and if anywhere else is tapped the media viewer will open. Tapping animated media will stop it.

## SDWebImage

SDWebImage comes with its own caching, prefetching, and decoding infrastructure; rather than maintain both, however, I have bridged Nuke and SDWebImage so that Nuke does the prefetching and SDWebImage does the decoding. This works as follows:
- The new `NukeWebpBridgeDecoder` overrides Nuke's default webp handling. It decodes the first frame as `ImageContainer.image`, checks whether the webp is animated, and, if so, bundles the raw data into `ImageContainer.data` (Nuke by default only decodes the first frame of a webp, and does not include the full data). 
- `MediaLoader` checks for the presence of the `data` in webp containers. If it detects data, it sets the image as `.webp` with the data; otherwise it sets it to `image`.
- `AnimatedImage` in `MediaView` takes the data, decodes it using SDWebImage tools (handled implicitly by `SDImageCodersManager.shared.addCoder(SDImageWebPCoder.shared)` in `MlemApp`), and renders it.

## Outstanding Work

- Performant mp4 thumbnails: NukeVideo doesn't give us a nice thumbnail to use for mp4s, so right now we just use a placeholder (like lemmy-ui).
- Better controls--right now it's just play/stop

### CODE REVIEW

File: Mlem/App/Views/Shared/Images/Core/Animated/VideoView.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+//  Created by Eric Andrews on 2024-09-23.
+//
+
+import AVFoundation
+import Foundation
+import Gifu
+import Nuke
+import NukeUI
+import NukeVideo
+import SwiftUI
```
Comment (Sjmarf) [inline Mlem/App/Views/Shared/Images/Core/Animated/VideoView.swift]: _nit_

```suggestion
import AVFoundation
import NukeVideo
import SwiftUI
```

File: Mlem.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 2]
```diff
@@ -28,13 +37,22 @@
...
+    },
     {
       "identity" : "mlemmiddleware",
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/mlemgroup/MlemMiddleware",
       "state" : {
-        "revision" : "1486eb4ed2834cc56f047535e42346bcbc2d5c19",
-        "version" : "0.38.0"
+        "branch" : "eric/gifs",
+        "revision" : "62e43a2e2552993984661016af26d7ee9b04d13a"
```
Comment (Sjmarf) [inline Mlem.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved]: Reminder

File: Mlem/App/Views/Shared/Images/Core/DynamicMediaView.swift [hunk 3]
```diff
@@ -61,22 +66,49 @@ struct DynamicImageView: View {
...
+        MediaView(media: loader.mediaType, playing: $playing)
             .overlay {
                 if showError, loader.error != nil {
                     errorOverlay
+                } else if loader.mediaType.isAnimated {
+                    if playing {
+                        Color.clear.contentShape(.rect)
+                            .onTapGesture {
+                                playing = false
+                            }
```
Comment (Sjmarf) [inline Mlem/App/Views/Shared/Images/Core/DynamicMediaView.swift:78]: This overrides the tap gesture to open the image viewer. Ideally we'd have a fullscreen button in the corner or something. Possibly OOS though

File: Mlem/App/Views/Shared/Images/Core/FixedImageView.swift [hunk 4]
```diff
@@ -14,7 +14,9 @@ import SwiftUI
 struct FixedImageView: View {
     @Environment(Palette.self) var palette
     
-    @State var loadingPref: ImageLoadingState? // tracked separately to allow correct propagation of inital value
+    @Setting(\.postSize) var postSize
+    
+    @State var loadingPref: MediaLoadingState? // tracked separately to allow correct propagation of inital value
```
Comment (Sjmarf) [inline Mlem/App/Views/Shared/Images/Core/FixedImageView.swift]:
```suggestion
    @State var loadingPref: MediaLoadingState? // tracked separately to allow correct propagation of initial value
```

File: Mlem/App/Views/Shared/Images/Core/MediaView.swift [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        case let .webp(_, animated):
+            AnimatedImage(data: animated)
+                .resizable()
+        default:
+            EmptyView()
+        }
+    }
+}
+
+extension UIImage {
```
Comment (Sjmarf) [inline Mlem/App/Views/Shared/Images/Core/MediaView.swift]: `private`?

```suggestion
private extension UIImage {
```

File: Mlem/App/Views/Shared/Images/Core/MediaView.swift [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+//
+
+import SDWebImageSwiftUI
+import SwiftUI
+import UIKit
+
+struct MediaView: View {
+    let media: MediaType
+    
+    @Binding var playing: Bool
```
Comment (Sjmarf) [inline Mlem/App/Views/Shared/Images/Core/MediaView.swift]: Doesn't need to be a `Binding` I don't think
```suggestion
    let playing: Bool
```

File: Mlem/App/Views/Shared/Images/Helpers/MediaLoader.swift [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            } else {
+                .image(.blank)
+            }
+        default:
+            .image(image)
+        }
+    }
+}
+
+func generateAVThumbnail(asset: AVAsset) -> UIImage {
```
Comment (Sjmarf) [inline Mlem/App/Views/Shared/Images/Helpers/MediaLoader.swift]: `private`?
```suggestion
private func generateAVThumbnail(asset: AVAsset) -> UIImage {
```

**Review by Sjmarf — APPROVED** [review]
Nice! Works really well 🔥 Some minor nits

File: Mlem/App/Views/Shared/Images/Core/DynamicMediaView.swift (hunk 3 above)
Comment (EricBAndrews) [reply Mlem/App/Views/Shared/Images/Core/DynamicMediaView.swift:78]: That is the intended behavior, at least for this iteration--UX has a lot of room to grow, leaving OOS.

File: Mlem/App/Views/Shared/Images/Core/MediaView.swift (hunk 6 above)
Comment (EricBAndrews) [reply Mlem/App/Views/Shared/Images/Core/MediaView.swift]: Good catch, holdover from an earlier iteration

**Pull request merged**